### ISSUE TITLE
docker + dat deployment

### ISSUE THREAD

**Issue by joehand** [datproject/dat#858]
Current status on docker + dat networking troubleshooting. This is a general issue to keep visibility and update progress, not related directly to cli.

We're still having trouble exposing ports directly. We had an issue with using IPTables before, which docker uses. So it may be related to that, not Docker directly.

I've tried exposing `3282, 3282/udp` directly via IPTables and wasn't able to get it to work without hole punching.

**Comment by joehand** [comment]
Seems like it is related to a bug in discovery-swarm/channel (https://github.com/mafintosh/discovery-swarm/issues/32) where the internal port doesn't match the external port and the wrong one gets reported.

**Comment by joehand** [comment]
Switching Docker networking to "host" mode works. This will use the OS ports instead of Docker's networking interface, getting around the external/internal port bug.

**Comment by mitar** [comment]
I think you are running it wrong. If you just do `-p 3282` then the host port will be randomly chosen. Which means that port dat inside will think it has will be different on the outside.

I am running hypercored inside Docker using `-p 3282:3282/tcp -p 3282:3282/udp` and it works perfectly.

**Comment by SvenDowideit** [comment]
interestingly, I've been having issues with dat sharing that are related to this:

```
docker run --rm -it -p 3282:3282/udp -p 3282:3282/tcp -v share_share:/data -v share_secretkeys:/root/.dat svendowideit/dat share /data
```

pretty much rarely works. I'm going to compare to running hypercored in a bit - cos this is ew!

**Comment by SvenDowideit** [comment]
I'm kinda wondering if there's not enough error reporting going on - I was able to run 2 `dat share`s on one computer - how are they both listening on the same ports?

**Comment by mitar** [comment]
@SvenDowideit see #945 about port question.

**Comment by SvenDowideit** [comment]
@mitar merci!

**Comment by mitar** [comment]
Also see #947.

**Comment by nettiopsu** [comment]
I have experienced the same thing while trying to dockerize dat. I tried exposing ports and with --net host as well, but it seems that host machine does not see it for some reason. dat doctor runs well with first tests without any errors, second test starts to use random ports and host machine fails to find it

Agree here with @SvenDowideit that it is almost impossible to understand where dat fails without proper debugging messages

**Comment by pfrazee** [comment]
Did you try running dat with the DEBUG env var set to *? That’ll give a lot of logging

**Comment by rjsteinert** [comment]
I'm facing the same issue, made worse by the fact I can't use `--net host` because my app relies on linked containers for the database. Watch out for the dreaded undefined behavior ;).

```
docker: Error response from daemon: conflicting options: host type networking can't be used with links. This would result in undefined behavior.
```

**Comment by maxogden** [comment]
I looked into this a while ago and found something about the docker NAT implementation having a bug with UDP hole punching. Can't find the github thread now but thought I'd comment in case that ends up being relevant.

**Comment by rjsteinert** [comment]
Thanks @maxogden I saw your status from two years ago to remind us two years in the future :-) https://twitter.com/maxogden/status/913874394421719040

This issue that @joehand references at the top may be the issue you are thinking of https://github.com/datproject/dat/issues/841

For the time being I'm mounting a volume and running `dat-party` (https://github.com/rjsteinert/dat-party) on the host machine so that any folder the container places in that volume will be shared as a Dat Archive from the host machine without having to punch through Docker NAT. It's a funny workaround.

**Comment by nettiopsu** [comment]
@pfrazee yes I tried, but it lacks still some data for understanding, sometimes it does not show even URL which dat is sharing

By the way, @rjsteinert reminded me about other Docker problem with dat share if you run it in MacOS. Docker has problems with sending notifications about file change (inotify) in this system. So if you run dat share for shared volume in Docker machine, then if you make changes in this volume from host machine, then dat will not get this changes neither share them in network. Though dat share works if changes were made from the same Docker machine, not from host. Taking this into account, it maybe wise for dat cli not to rely always on inotify, but also run sometimes in low priority scan of all files in folder for changes (?) at least for Docker

**Comment by thadguidry** [comment]
@mitar but it does ?  It picks first available, if the default port is already in use:  https://github.com/datproject/dat/blob/master/bin/cli.js#L39

**Comment by mitar** [comment]
Yes, and so it is tricky to know that default port is not being used. So it is tricky to debug. I would prefer a mode which would be "use given port or die".

i have had issues in Docker only because my dat daemon was listening on ports I have not had mapped. And it was hard to debug.

**Comment by joehand** [comment]
> Yes, and so it is tricky to know that default port is not being used. So it is tricky to debug. I would prefer a mode which would be "use given port or die".

@mitar, the latest release of the dat cli fixed this =). Thanks for the suggestions on that.

**Comment by mitar** [comment]
Then it should work. (If nothing else changed. I have not used dat for few months now.)

> I tried exposing port 3282

@nettiopsu, did you expose both UDP and TCP ports? Use `-p 3282:3282/tcp -p 3282:3282/udp`.

**Comment by nettiopsu** [comment]
@mitar just tried, it did not help, same situation - can not connect from host machine, but can connect from a remote one

I have tried also this new port parameter to use some another port, but then even remote machine stopped working

**Comment by mitar** [comment]
Oh, yes, from host it does not work. This is a general issue with NAT.

**Comment by drelephant** [comment]
Would you mind explaining more about what the general issue with NAT is?

**Comment by mitar** [comment]
I mean, it depends how your Docker is configured, but if whole Docker network is behind one IP (NAT), you might not be able to access a particular IP of the container from outside (host). You can map port to that one IP and then access the port on it.
